Quarkus version of jbang does not support M1, but mainline jbang now does

### Describe the bug

I'm seeing some test failures related to M1 support. Looking into them a bit, they're originating in the `jbang` script. The good news is that jbang has added some fixes for M1 which we can pull across. 

I can see that https://github.com/jbangdev/jbang/issues/650 added partial support, to avoid failing in cases where an M1 JVM was present on the system. https://github.com/jbangdev/jbang/pull/1269 is more recent and will do a download, but force java17 (since that's all Temurium has).  



### Expected behavior

`quarkus build` on M1 (without a native build) should build cleanly. 

### Actual behavior

The build fails on M1: 

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CliProjectJBangTest.testCreateAppDefaults:64 Expected OK return code. Result:
result: {
  exitCode: {1},
  system_err: {Unsupported Architecture: arm64
},
  system_out: {}
} ==> expected: <0> but was: <1>
[ERROR]   CliProjectJBangTest.testCreateAppOverrides:98 Expected OK return code. Result:
result: {
  exitCode: {1},
  system_err: {Unsupported Architecture: arm64
},
  system_out: {}
} ==> expected: <0> but was: <1>
[ERROR]   CliProjectJBangTest.testCreateCliDefaults:127 Expected OK return code. Result:
result: {
  exitCode: {1},
  system_err: {Unsupported Architecture: arm64
},
  system_out: {[DEBUG] Build project with initial parameters: Build [buildOptions=BuildOptions [buildNative=false, clean=true, offline=false, skipTests=false], properties={property=value1, property2=value2, maven.repo.local=/Users/holly/.m2/repository, maven.settings=/Users/holly/.m2/settings.xml}, output=OutputOptions [testMode=true, showErrors=true, verbose=true], params=[]]
/Users/holly/Code/quarkus/myfork/quarkus/devtools/cli/target/test-project/CliProjectJBangTest/code-with-quarkus/jbang --verbose --fresh build -Dproperty=value1 -Dproperty2=value2 -Dmaven.repo.local=/Users/holly/.m2/repository -Dmaven.settings=/Users/holly/.m2/settings.xml -DquarkusRegistryClient=true src/main.java

}
} ==> expected: <0> but was: <1>
[INFO] 
[ERROR] Tests run: 78, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[WARNING] Corrupted STDOUT by directly writing to native stream in forked JVM 1. See FAQ web page and the dump file /Users/holly/Code/quarkus/myfork/quarkus/devtools/cli/target/surefire-reports/2022-04-26T18-03-47_990-jvmRun1.dumpstream
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:19 min
[INFO] Finished at: 2022-04-26T18:05:03+01:00
```

Running the jbang build directly also fails:

```Unsupported Architecture: arm64```

### How to Reproduce?

This can be seen by running a build with tests on an M1 system: 

1. ```./mvnw -Dquickly -DskipTests=false -f devtools/cli ```

Alternatively, the jbang build command can be run directly, and will fail: 

```
holly@hcummins-mac quarkus % quarkus build -e -B --clean -Dproperty=value1 -Dproperty2=value2 -Dmaven.repo.local=/Users/holly/.m2/repository --cli-test --cli-test-dir /Users/holly/Code/quarkus/myfork/quarkus/devtools/cli/target/test-project/CliProjectJBangTest/cli-nested/my-project
Unsupported Architecture: arm64
```

### Output of `uname -a` or `ver`

Darwin hcummins-mac 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.2" 2022-01-18 OpenJDK Runtime Environment Temurin-17.0.2+8 (build 17.0.2+8) OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (build 17.0.2+8, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

HEAD

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven home: /Users/holly/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4 Java version: 17.0.2, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home Default locale: en_GB, platform encoding: UTF-8 OS name: "mac os x", version: "12.3.1", arch: "aarch64", family: "mac"

### Additional information

_No response_

--- a/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus-jbang/tooling/jbang-wrapper/base/jbang
+++ b/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus-jbang/tooling/jbang-wrapper/base/jbang
@@ -2,7 +2,7 @@
 
 #
 # To run this script remotely type this in your shell
-# (where <args>... are the arguments you want to pass to Jbang):
+# (where <args>... are the arguments you want to pass to JBang):
 #   curl -Ls https://sh.jbang.dev | bash -s - <args>...
 #
 
@@ -17,21 +17,21 @@ absolute_path() {
 
 resolve_symlink() {
   if [[ $OSTYPE != darwin* ]]; then minusFarg="-f"; fi
-  sym_resolved=$(readlink ${minusFarg} $1)
+  sym_resolved=$(readlink ${minusFarg} "$1")
 
   if [[ -n $sym_resolved ]]; then
-    echo $sym_resolved
+    echo "$sym_resolved"
   else
-    echo $1
+    echo "$1"
   fi
 }
 
 download() {
   if [ -x "$(command -v curl)" ]; then
-    curl -sLf -H "Accept: application/gzip, application/octet-stream" -o "$2" $1
+    curl -sLf -o "$2" "$1"
     retval=$?
   elif [ -x "$(command -v wget)" ]; then
-    wget -q --header="Accept: application/gzip, application/octet-stream" -O "$2" $1
+    wget -q -O "$2" "$1"
     retval=$?
   else
     echo "Error: curl or wget not found, please make sure one of them is installed" 1>&2
@@ -39,16 +39,52 @@ download() {
   fi
 }
 
-abs_jbang_path=$(resolve_symlink $(absolute_path $0))
+unpack() {
+  if [[ "$1" == *.tar.gz ]]; then
+    gzip -cd "$1" | tar xf - -C "$2"
+    retval=$?
+    if [[ $os == mac && $retval -eq 0 ]]; then
+      mv "$TDIR/jdks/$javaVersion.tmp/"*/Contents/Home/* "$TDIR/jdks/$javaVersion.tmp/"
+      retval=$?
+    else
+      mv "$TDIR/jdks/$javaVersion.tmp/"*/* "$TDIR/jdks/$javaVersion.tmp/"
+    fi
+  else
+    unzip -qq -o "$1" -d "$2"
+    retval=$?
+    mv "$TDIR/jdks/$javaVersion.tmp/"*/* "$TDIR/jdks/$javaVersion.tmp/"
+  fi
+}
+
+javacInPath() {
+  [[ -x "$(command -v javac)" && ( $os != "mac" || $(/usr/libexec/java_home &> /dev/null) ) ]]
+}
+
+abs_jbang_dir=$(dirname "$(resolve_symlink "$(absolute_path "$0")")")
+
+# todo might vary by os so can be overwritten below
+libc_type=glibc
+file_type=tar.gz
 
 case "$(uname -s)" in
   Linux*)
-    os=linux;;
+    os=linux
+    if [ -f /etc/alpine-release ]; then
+      os=alpine-linux
+      javaVersion=16
+    fi
+    ;;
   Darwin*)
-    os=mac;;
+    os=mac
+    libc_type=libc;;
   CYGWIN*|MINGW*|MSYS*)
-    os=windows;;
-  *)        echo "Unsupported Operating System: $(uname -s)" 1>&2; exit 1;;
+    os=windows
+    libc_type=c_std_lib
+    file_type=zip;;
+  AIX)
+    os=aix;;
+  *)
+    os=
 esac
 
 case "$(uname -m)" in
@@ -60,26 +96,55 @@ case "$(uname -m)" in
     arch=aarch64;;
   armv7l)
     arch=arm;;
+  ppc64le)
+    arch=ppc64le;;
+  s390x)
+    arch=s390x;;
+  arm64)
+    if [ "$os" = "mac" ]; then
+      arch=arm64
+      ## force use of 17 as 11 not yet available https://github.com/adoptium/adoptium/issues/96
+      javaVersion=17
+    else
+      arch=arm64
+    fi
+    ;;
   *)
-    echo "Unsupported Architecture: $(uname -m)" 1>&2; exit 1;;
+    ## AIX gives a machine ID for `uname -m` but it only supports ppc64
+    if [ "$os" = "aix" ]; then
+      arch=ppc64
+    else
+      arch=
+    fi
+    ;;
 esac
 
-## when mingw/git bash or cygwin fall out to just running the bat file.
-if [[ $os == windows ]]; then
-  $(dirname $abs_jbang_path)/jbang.cmd $*
+## when using cygwin fall out to just running the bat file.
+if [[ $os == windows && -f "$abs_jbang_dir/jbang.cmd" && "$(uname -s)" == CYGWIN* ]]; then
+  cmd /c "$(cygpath -m "$abs_jbang_dir"/jbang.cmd)" "$@"
   exit $?
 fi
+
+if [[ -z "$JBANG_JDK_VENDOR" ]]; then
+  if [[ "$javaVersion" -eq 8 || "$javaVersion" -eq 11 || "$javaVersion" -ge 17 ]]; then
+    distro="temurin";
+  else
+    distro="aoj"; fi
+else
+  distro=$JBANG_JDK_VENDOR
+fi
+if [[ -z "$JBANG_JDK_RELEASE" ]]; then release="ga"; else release="$JBANG_JDK_RELEASE"; fi
 
 if [[ -z "$JBANG_DIR" ]]; then JBDIR="$HOME/.jbang"; else JBDIR="$JBANG_DIR"; fi
 if [[ -z "$JBANG_CACHE_DIR" ]]; then TDIR="$JBDIR/cache"; else TDIR="$JBANG_CACHE_DIR"; fi
 
 ## resolve application jar path from script location
-if [ -f "$(dirname $abs_jbang_path)/jbang.jar" ]; then
-  jarPath=$(dirname $abs_jbang_path)/jbang.jar
-elif [ -f "$(dirname $abs_jbang_path)/.jbang/jbang.jar" ]; then
-  jarPath=$(dirname $abs_jbang_path)/.jbang/jbang.jar
+if [ -f "$abs_jbang_dir/jbang.jar" ]; then
+  jarPath=$abs_jbang_dir/jbang.jar
+elif [ -f "$abs_jbang_dir/.jbang/jbang.jar" ]; then
+  jarPath=$abs_jbang_dir/.jbang/jbang.jar
 else
-  if [ ! -f "$JBDIR/bin/jbang.jar" ]; then
+  if [[ ! -f "$JBDIR/bin/jbang.jar" || ! -f "$JBDIR/bin/jbang" ]]; then
     echo "Downloading JBang..." 1>&2
     mkdir -p "$TDIR/urls"
     jburl="https://github.com/jbangdev/jbang/releases/latest/download/jbang.tar"
@@ -93,46 +158,55 @@ else
     rm -f "$JBDIR/bin/jbang" "$JBDIR/bin"/jbang.*
     cp -f "$TDIR/urls/jbang/bin"/* "$JBDIR/bin"
   fi
-  eval "exec $JBDIR/bin/jbang $*"
+  "$JBDIR"/bin/jbang "$@"
+  exit $?
+fi
+if [ -f "$jarPath.new" ]; then
+  # a new jbang version was found, we replace the old one with it
+  mv "$jarPath.new" "$jarPath"
 fi
 
 # Find/get a JDK
 unset JAVA_EXEC
 if [[ -n "$JAVA_HOME" ]]; then
   # Determine if a (working) JDK is available in JAVA_HOME
-  if [ -x "$(command -v $JAVA_HOME/bin/javac)" ]; then
-    JAVA_EXEC="$JAVA_HOME/bin/java";
+  if [ -x "$(command -v "$JAVA_HOME"/bin/javac)" ]; then
+    JAVA_EXEC="$JAVA_HOME/bin/java"
   else
     echo "JAVA_HOME is set but does not seem to point to a valid Java JDK" 1>&2
   fi
 fi
 if [[ -z "$JAVA_EXEC" ]]; then
   # Determine if a (working) JDK is available on the PATH
-  if [ -x "$(command -v javac)" ]; then
-    JAVA_EXEC="java";
+  if javacInPath; then
+    unset JAVA_HOME
+    JAVA_EXEC="java"
   elif [ -x "$JBDIR/currentjdk/bin/javac" ]; then
     export JAVA_HOME="$JBDIR/currentjdk"
-    JAVA_EXEC="$JBDIR/currentjdk/bin/java";
+    JAVA_EXEC="$JBDIR/currentjdk/bin/java"
   else
     export JAVA_HOME="$TDIR/jdks/$javaVersion"
     JAVA_EXEC="$JAVA_HOME/bin/java"
     # Check if we installed a JDK before
     if [ ! -d "$TDIR/jdks/$javaVersion" ]; then
       # If not, download and install it
+      if [[ $os == "" ]]; then
+        echo "Unable to download JDK, unsupported Operating System: $(uname -s)" 1>&2
+        exit 1
+      fi
+      if [[ $arch == "" ]]; then
+        echo "Unable to download JDK, unsupported Architecture: $(uname -m)" 1>&2
+        exit 1
+      fi
       mkdir -p "$TDIR/jdks"
       echo "Downloading JDK $javaVersion. Be patient, this can take several minutes..." 1>&2
-      jdkurl="https://api.adoptopenjdk.net/v3/binary/latest/$javaVersion/ga/$os/$arch/jdk/hotspot/normal/adoptopenjdk"
-      download $jdkurl "$TDIR/bootstrap-jdk.tgz"
+      jdkurl="https://api.foojay.io/disco/v2.0/directuris?distro=$distro&javafx_bundled=false&libc_type=$libc_type&archive_type=$file_type&operating_system=$os&package_type=jdk&version=$javaVersion&release_status=$release&architecture=$arch&latest=available"
+      download "$jdkurl" "$TDIR/bootstrap-jdk.$file_type"
       if [ $retval -ne 0 ]; then echo "Error downloading JDK" 1>&2; exit $retval; fi
       echo "Installing JDK $javaVersion..." 1>&2
       rm -rf "$TDIR/jdks/$javaVersion.tmp/"
       mkdir -p "$TDIR/jdks/$javaVersion.tmp"
-      tar xf "$TDIR/bootstrap-jdk.tgz" -C "$TDIR/jdks/$javaVersion.tmp" --strip-components=1
-      retval=$?
-      if [[ $os == mac && $retval -eq 0 ]]; then
-        mv "$TDIR/jdks/$javaVersion.tmp/Contents/Home/"* "$TDIR/jdks/$javaVersion.tmp/"
-        retval=$?
-      fi
+      unpack "$TDIR/bootstrap-jdk.$file_type" "$TDIR/jdks/$javaVersion.tmp"
       if [ $retval -ne 0 ]; then
         # Check if the JDK was installed properly
         javac -version > /dev/null 2>&1
@@ -142,7 +216,7 @@ if [[ -z "$JAVA_EXEC" ]]; then
       # Activate the downloaded JDK giving it its proper name
       mv "$TDIR/jdks/$javaVersion.tmp" "$TDIR/jdks/$javaVersion"
       # Set the current JDK
-      ${JAVA_EXEC} -classpath ${jarPath} dev.jbang.Main jdk default $javaVersion
+      ${JAVA_EXEC} -classpath "${jarPath}" dev.jbang.Main jdk default $javaVersion
     fi
   fi
 fi
@@ -151,11 +225,13 @@ fi
 ## attempt to ensure each argument keeps its original quoting
 
 ## run it using command substitution to have just the user process once jbang is done
-output=$(CLICOLOR_FORCE=1 ${JAVA_EXEC} ${JBANG_JAVA_OPTIONS} -classpath ${jarPath} dev.jbang.Main "$@")
+export JBANG_RUNTIME_SHELL=bash
+export JBANG_STDIN_NOTTY=$([ -t 0 ] && echo "false" || echo "true")
+output=$(CLICOLOR_FORCE=1 ${JAVA_EXEC} ${JBANG_JAVA_OPTIONS} -classpath "${jarPath}" dev.jbang.Main "$@")
 err=$?
 if [ $err -eq 255 ]; then
   eval "exec $output"
-elif [ ! -z "$output" ]; then
+elif [ -n "$output" ]; then
   echo "$output"
   exit $err
 else

--- a/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus-jbang/tooling/jbang-wrapper/base/jbang.cmd
+++ b/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus-jbang/tooling/jbang-wrapper/base/jbang.cmd
@@ -1,15 +1,8 @@
 @echo off
 SETLOCAL ENABLEEXTENSIONS ENABLEDELAYEDEXPANSION
 
-
 rem The Java version to install when it's not installed on the system yet
 if "%JBANG_DEFAULT_JAVA_VERSION%"=="" (set javaVersion=11) else (set javaVersion=%JBANG_DEFAULT_JAVA_VERSION%)
-
-set os=windows
-set arch=x64
-
-set jburl="https://github.com/jbangdev/jbang/releases/latest/download/jbang.zip"
-set jdkurl="https://api.adoptopenjdk.net/v3/binary/latest/%javaVersion%/ga/%os%/%arch%/jdk/hotspot/normal/adoptopenjdk"
 
 if "%JBANG_DIR%"=="" (set JBDIR=%userprofile%\.jbang) else (set JBDIR=%JBANG_DIR%)
 if "%JBANG_CACHE_DIR%"=="" (set TDIR=%JBDIR%\cache) else (set TDIR=%JBANG_CACHE_DIR%)
@@ -21,20 +14,16 @@ if exist "%~dp0jbang.jar" (
   set jarPath=%~dp0.jbang\jbang.jar
 ) else (
   if not exist "%JBDIR%\bin\jbang.jar" (
-    echo Downloading JBang... 1>&2
-    if not exist "%TDIR%\urls" ( mkdir "%TDIR%\urls" )
-    powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "$ProgressPreference = 'SilentlyContinue'; Invoke-WebRequest %jburl% -OutFile %TDIR%\urls\jbang.zip"
-    if !ERRORLEVEL! NEQ 0 ( echo Error downloading JBang 1>&2 & exit /b %ERRORLEVEL% )
-    echo Installing JBang... 1>&2
-    if exist "%TDIR%\urls\jbang" ( rd /s /q "%TDIR%\urls\jbang" > nul 2>&1 )
-    powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "$ProgressPreference = 'SilentlyContinue'; Expand-Archive -Path %TDIR%\urls\jbang.zip -DestinationPath %TDIR%\urls"
-    if !ERRORLEVEL! NEQ 0 ( echo Error installing JBang 1>&2 & exit /b %ERRORLEVEL% )
-    if not exist "%JBDIR%\bin" ( mkdir "%JBDIR%\bin" )
-    del /f /q "%JBDIR%\bin\jbang" "%JBDIR%\bin\jbang.*"
-    copy /y "%TDIR%\urls\jbang\bin\*" "%JBDIR%\bin" > nul 2>&1
+    powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "%~dp0jbang.ps1 version" > nul
+    if !ERRORLEVEL! NEQ 0 ( exit /b %ERRORLEVEL% )
   )
   call "%JBDIR%\bin\jbang.cmd" %*
   exit /b %ERRORLEVEL%
+)
+if exist "%jarPath%.new" (
+  rem a new jbang version was found, we replace the old one with it
+  copy /y "%jarPath%.new" "%jarPath%" > nul 2>&1
+  del /f /q "%jarPath%.new"
 )
 
 rem Find/get a JDK
@@ -51,6 +40,7 @@ if "!JAVA_EXEC!"=="" (
   rem Determine if a (working) JDK is available on the PATH
   where javac > nul 2>&1
   if !errorlevel! equ 0 (
+    set JAVA_HOME=
     set JAVA_EXEC=java.exe
   ) else if exist "%JBDIR%\currentjdk\bin\javac" (
     set JAVA_HOME=%JBDIR%\currentjdk
@@ -61,33 +51,19 @@ if "!JAVA_EXEC!"=="" (
     rem Check if we installed a JDK before
     if not exist "%TDIR%\jdks\%javaVersion%" (
       rem If not, download and install it
-      if not exist "%TDIR%\jdks" ( mkdir "%TDIR%\jdks" )
-      echo Downloading JDK %javaVersion%. Be patient, this can take several minutes... 1>&2
-      powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "$ProgressPreference = 'SilentlyContinue'; Invoke-WebRequest %jdkurl% -OutFile %TDIR%\bootstrap-jdk.zip"
-      if !ERRORLEVEL! NEQ 0 ( echo Error downloading JDK 1>&2 & exit /b %ERRORLEVEL% )
-      echo Installing JDK %javaVersion%... 1>&2
-      if exist "%TDIR%\jdks\%javaVersion%.tmp" ( rd /s /q "%TDIR%\jdks\%javaVersion%.tmp" > nul 2>&1 )
-      powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "$ProgressPreference = 'SilentlyContinue'; Expand-Archive -Path %TDIR%\bootstrap-jdk.zip -DestinationPath %TDIR%\jdks\%javaVersion%.tmp"
-      if !ERRORLEVEL! NEQ 0 ( echo Error installing JDK 1>&2 & exit /b %ERRORLEVEL% )
-      for /d %%d in (%TDIR%\jdks\%javaVersion%.tmp\*) do (
-        powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "Move-Item %%d\* !TDIR!\jdks\%javaVersion%.tmp"
-        if !ERRORLEVEL! NEQ 0 ( echo Error installing JDK 1>&2 & exit /b %ERRORLEVEL% )
-      )
-      rem Check if the JDK was installed properly
-      %TDIR%\jdks\%javaVersion%.tmp\bin\javac -version > nul 2>&1
-      if !ERRORLEVEL! NEQ 0 ( echo "Error installing JDK" 1>&2; exit /b %ERRORLEVEL% )
-      rem Activate the downloaded JDK giving it its proper name
-      ren "%TDIR%\jdks\%javaVersion%.tmp" "%javaVersion%"
+      echo powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command "%~dp0jbang.ps1 jdk install %JBANG_DEFAULT_JAVA_VERSION%"
+      if !ERRORLEVEL! NEQ 0 ( exit /b %ERRORLEVEL% )
+      rem Set the current JDK
+      !JAVA_EXEC! -classpath "%jarPath%" dev.jbang.Main jdk default "%javaVersion%"
     )
-    # Set the current JDK
-    !JAVA_EXEC! -classpath "%jarPath%" dev.jbang.Main jdk default "%javaVersion%"
   )
 )
 
 if not exist "%TDIR%" ( mkdir "%TDIR%" )
 set tmpfile=%TDIR%\%RANDOM%.jbang.tmp
 rem execute jbang and pipe to temporary random file
-set JBANG_USES_POWERSHELL=
+set JBANG_RUNTIME_SHELL=cmd
+2>nul >nul timeout /t 0 && (set JBANG_STDIN_NOTTY=false) || (set JBANG_STDIN_NOTTY=true)
 set "CMD=!JAVA_EXEC!"
 SETLOCAL DISABLEDELAYEDEXPANSION
 %CMD% > "%tmpfile%" %JBANG_JAVA_OPTIONS% -classpath "%jarPath%" dev.jbang.Main %*

